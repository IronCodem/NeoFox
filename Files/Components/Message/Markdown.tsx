import React from 'react';
import styles from './../../Style/Components/Markdown.module.scss';
import Markdown from '../../Scripts/Markdown_Standard';
import Prism, { Token, TokenStream } from 'prismjs';
import hljs from '../../Scripts/hljs';
import { Image } from './Preview';

import Chatter_User from '../../Scripts/Chatter';
// Function
type MyProps = { MD: string, User: Chatter_User };
let Marked = (props: MyProps) => {
  let Tokens: TokenStream = Prism.tokenize(
    props.MD, 
    Markdown
  );
  let Output: Jsx.Element[] = [];
  Tokens.forEach((Token: Token | string, i) => {
    if (!Token) return;
    let Convert = ({ content }: any) => {
      if (typeof content == 'string') return { Text: content, Type: '' };
      let Text: string = '', Output: any = { language: '', Type: '', Url: '' };
      content.forEach((Part: any) => {
        if (typeof Part == 'string') {
          Text += Part;
        } else if (Part.type == 'language') {
          Output['language'] = Part.content.trim();
        } else if (Part.type == 'Type') {
          Output['Type'] = Part.content;
        } else if (Part.type == 'Url') {
          Output['Url'] = Part.content;
        }
      });
      return { Text, Output };
    }
    if (typeof Token == 'string' && Token.trim()) {
      Output.push(<p key={i}>{Token}</p>);
    } else {
      //@ts-ignore
      switch(Token.type) {
        case 'bold':
          Output.push(<strong key={i}>{Convert(Token).Text}</strong>);
          break;
        case 'italic':
          Output.push(<em key={i}>{Convert(Token).Text}</em>);
          break;
        case 'bolditalic':
          Output.push(<strong key={i}><em>{Convert(Token).Text}</em></strong>);
          break;
        case 'strikethrough':
          Output.push(<del key={i}>{Convert(Token).Text}</del>);
          break;
        case 'underline':
          Output.push(<ins key={i}>{Convert(Token).Text}</ins>);
          break;
        case 'code':
          Output.push(<code key={i}>{Convert(Token).Text}</code>);
          break;
        case 'blockquote':
          Output.push(<blockquote key={i} className={styles.blockqoute}>{Convert(Token).Text}</blockquote>);
          break;
        case 'url':
          let { Type:Title, Url:Href } = Convert(Token).Output;
          try {
            Href = encodeURI(Href.replace('javascript:','')).replace(/%25/g, '%');
          } catch (e) { Href = ''; }
          if (Href == '') Output.push(<p key={i}>{Title}</p>);
          else Output.push(<a key={i} href={Href}>{Title||Href}</a>);
          break;
        case 'url-reference':
          let { Type, Url } = Convert(Token).Output;
          try {
            Url = encodeURI(Url.replace('javascript:','')).replace(/%25/g, '%');
          } catch (e) { Url = ''; }
          if (Url == '') Output.push(<p key={i}>{Type}</p>);
          else {
            switch (true) {
              case /\.(gif|jpe?g|tiff?|png|webp|bmp)$/i.test(Url) || Type == 'Image':
              default:
                Output.push(
                  <Image
                    key={i}
                    Url={Url}
                    alt={Type || ''}
                  />
                );
            }
          }
          break;
        case 'code_block':
          let Data = Convert(Token), Txt: string = Data.Text.trim();
          let Highlighted = hljs.getLanguage(Data.Output.language)?hljs.highlight(Txt, { language: Data.Output.language }).value:Txt;
          Output.push(
            <pre key={i}><code dangerouslySetInnerHTML={{ __html: Highlighted }}></code></pre>
          );
          break;
        case 'h1':
          Output.push(<h1 key={i}>{Convert(Token).Text}</h1>);
          break;
        case 'h2':
          Output.push(<h2 key={i}>{Convert(Token).Text}</h2>);
          break;
        case 'h3':
          Output.push(<h3 key={i}>{Convert(Token).Text}</h3>);
          break;
        case 'channel':
          Output.push(<span key={i} className={styles.channel} onClick={() => props.User.ActiveChannelName(Convert(Token).Text)}>#{Convert(Token).Text}</span>);
          break;
        case 'orderedlist':
          Output.push(
            <ol key={i}>
            {Convert(Token).Text.split('\n').map((Text, n) => <li key={n}>{Text}</li>)}
            </ol>
          );
          break;
        case 'unorderedlist':
          Output.push(
            <ul key={i}>
            {Convert(Token).Text.split('\n').map((Text, n) => <li key={n}>{Text}</li>)}
            </ul>
          );
          break;
        case 'horizontalRule':
          Output.push(<hr key={i} />);
          break;
      }
    }
  })
  return (<>{Output}</>);
}


export default Marked;