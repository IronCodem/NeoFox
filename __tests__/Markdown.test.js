const React = require('react');
const renderer = require('react-test-renderer');
import Markdown from './../Files/Components/Message/Markdown';

// ========
// Markdown
// ========
test('Markdown--Attachment', async function () {
  const component = renderer.create(<Markdown MD={'![Image](htttps://repl.it/)'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Url', async function () {
  const component = renderer.create(<Markdown MD={'[Repl](htttps://repl.it/)'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Url-2', async function () {
  const component = renderer.create(<Markdown MD={'<htttps://repl.it/>'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--BlockQoute', async function () {
  const component = renderer.create(<Markdown MD={'> Test'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--InlineCode', async function () {
  const component = renderer.create(<Markdown MD={'`Inline`'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Codeblock', async function () {
  const component = renderer.create(<Markdown MD={'```js\nconsole.log(\'Hello world\');\n```'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Bold', async function () {
  const component = renderer.create(<Markdown MD={'**Bold**'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--StrikeThrough', async function () {
  const component = renderer.create(<Markdown MD={'~~StrikeThrough'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Underline', async function () {
  const component = renderer.create(<Markdown MD={'__Underline__'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Italic', async function () {
  const component = renderer.create(<Markdown MD={'*console.log(\'Hello world\');*'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Bold--And--Italic', async function () {
  const component = renderer.create(<Markdown MD={'***test***'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-1', async function () {
  const component = renderer.create(<Markdown MD={'# h1'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-1-Type-2', async function () {
  const component = renderer.create(<Markdown MD={'h1\n===='} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-2', async function () {
  const component = renderer.create(<Markdown MD={'## h2'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-2-Type-2', async function () {
  const component = renderer.create(<Markdown MD={'h2\n---'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-3', async function () {
  const component = renderer.create(<Markdown MD={'### h3'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-4', async function () {
  const component = renderer.create(<Markdown MD={'#### h4'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-5', async function () {
  const component = renderer.create(<Markdown MD={'##### h5'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--Header-6', async function () {
  const component = renderer.create(<Markdown MD={'###### h6'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Markdown--Channel', async function () {
  const component = renderer.create(<Markdown MD={'#Main'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// Markdown Lists
test('Markdown--Ordered--List', async function () {
  const component = renderer.create(<Markdown MD={'1. test\n2. test'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--unordered--List', async function () {
  const component = renderer.create(<Markdown MD={'* test\n* test'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--unordered--List-2', async function () {
  const component = renderer.create(<Markdown MD={'+ test\n+ test'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('Markdown--unordered--List-3', async function () {
  const component = renderer.create(<Markdown MD={'- test\n- test'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});


test('Markdown--Horizontal--Rule', async function () {
  const component = renderer.create(<Markdown MD={'test\n\n---\n\n'} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});