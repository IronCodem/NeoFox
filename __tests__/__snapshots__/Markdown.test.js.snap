// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markdown--Attachment 1`] = `
<picture
  onClick={[Function]}
>
  <img
    alt="Image"
    height="500"
    src="htttps://repl.it/"
  />
</picture>
`;

exports[`Markdown--BlockQoute 1`] = `
<blockquote>
  Test
</blockquote>
`;

exports[`Markdown--Bold 1`] = `
<strong>
  Bold
</strong>
`;

exports[`Markdown--Bold--And--Italic 1`] = `
<strong>
  <em>
    test
  </em>
</strong>
`;

exports[`Markdown--Channel 1`] = `
<span
  onClick={[Function]}
>
  #
  Main
</span>
`;

exports[`Markdown--Codeblock 1`] = `
<pre>
  <code
    dangerouslySetInnerHTML={
      Object {
        "__html": "<span class=\\"hljs-built_in\\">console</span>.log(<span class=\\"hljs-string\\">&#x27;Hello world&#x27;</span>);",
      }
    }
  />
</pre>
`;

exports[`Markdown--Header-1 1`] = `
<h1>
  h1
</h1>
`;

exports[`Markdown--Header-1-Type-2 1`] = `
<h1>
  h1
</h1>
`;

exports[`Markdown--Header-2 1`] = `
<h2>
  h2
</h2>
`;

exports[`Markdown--Header-2-Type-2 1`] = `
<h2>
  h2
</h2>
`;

exports[`Markdown--Header-3 1`] = `
<h3>
  h3
</h3>
`;

exports[`Markdown--Header-4 1`] = `null`;

exports[`Markdown--Header-5 1`] = `null`;

exports[`Markdown--Header-6 1`] = `null`;

exports[`Markdown--Horizontal--Rule 1`] = `
Array [
  <p>
    test
  </p>,
  <hr />,
]
`;

exports[`Markdown--InlineCode 1`] = `
<code>
  Inline
</code>
`;

exports[`Markdown--Italic 1`] = `
<em>
  console.log('Hello world');
</em>
`;

exports[`Markdown--Ordered--List 1`] = `
<ol>
  <li>
    test
  </li>
  <li>
    test
  </li>
</ol>
`;

exports[`Markdown--StrikeThrough 1`] = `
<p>
  ~~StrikeThrough
</p>
`;

exports[`Markdown--Underline 1`] = `
<ins>
  Underline
</ins>
`;

exports[`Markdown--Url 1`] = `
<a
  href="htttps://repl.it/"
>
  Repl
</a>
`;

exports[`Markdown--Url-2 1`] = `
<a
  href="htttps://repl.it/"
>
  htttps://repl.it/
</a>
`;

exports[`Markdown--unordered--List 1`] = `
<ul>
  <li>
    test
  </li>
  <li>
    test
  </li>
</ul>
`;

exports[`Markdown--unordered--List-2 1`] = `
<ul>
  <li>
    test
  </li>
  <li>
    test
  </li>
</ul>
`;

exports[`Markdown--unordered--List-3 1`] = `
<ul>
  <li>
    test
  </li>
  <li>
    test
  </li>
</ul>
`;
